### DISCUSSION TITLE
whdload support for demos which require real kickstarts

### DISCUSSION THREAD

**Discussion by mithrendal** [vAmigaWeb/vAmigaWeb#219] [Ideas]
@wepl 

For demos which require a commodore kickstart you could do the following:

1. Display a locked symbol when no proper kickstart found in browsers local database 

![IMG_0101](https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/7140f594-2fc6-4ee6-924f-0ba5623acc7c)

   (locked - kickstart 3.1 required, to view drag kickstart file into here)

2.  When user drags the kickstart file on the lock symbol => save it to local browser database and show the play symbol

3. When the required Commodore kickstart was found in local browser database show play symbol 

When user clicks on play, do read kickstart from local browser database and insert into vAmigaWeb_players file descriptor.

If you think this way is interesting for you, we could support and help you in achieving/doing it.

BTW: we did something similar in another web port of vAmiga.   https://vamiganet.github.io  (click on showcases there are some titles which require commodore roms)

![IMG_0102](https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/7e1a539f-63c6-4774-acc9-94a211660849)

**Comment by dirkwhoffmann** [comment]
@mithrendal: I think it's best to integrate the lock-button functionality directly in the Player. 

@wepl: What do you think about this functionality? It would be an efficient approach to letting users run stuff that requires an original Kickstart. All he would have to do is read the note under the lock symbol and drag an original Rom onto the lock icon.

**Comment by wepl** [comment]
>> When user clicks on play, do read kickstart from local browser database and insert into vAmigaWeb_players file descriptor.

>How does this work?

Modern browsers have a thing called `indexed-db`. This is a local database where the web site e.g. whdload.de can store data on the users local browser space for later use (opposed to upload to a server). We can use it for whdload.de i.e. the user selects the kickstart and with some javascript commands we can store the kickstart into the indexed-db. 

I think what Dirk said is a good way to implement that feature. I.e. we put the necessary javascript code inside the vAmigaWeb-player component.   

The website would just had to express the requirement e.g. like

```javascript
vAmigaWeb_player.samesite_file=
{
  kickstart_rom_required:  [2.0, 3.1],
  url:'/va/ctros/UnitA_FA18Interceptor.zip',
  name:'UnitA_FA18Interceptor.zip'
};
```
 
The player then would check the local db for the kickstart, 
- case kickstart is found in indexed-db: 
    it will show the play icon
    on click it would load it and hand the rom over to the emulator.

- case kickstart is not found: 
    player shows the lock icon
    it will ask the user to select one and store it
    when selected file is valid => show play button (or alternatively immediately load vAmigaWeb) 

This functionality would be done in the player component… we need a new function to do the rendering of the lock vs play icon.

I.e.

Before we call
vAmigaWeb_player.load(…)

We need to call another method before, i.e. when the page is loads … 

Maybe something like 

vAmigaWeb_player.setup(…)

Install will do the rendering of the lock or play icon
Does anyone has a better name for it ?

>I would need access to the kickstart (file) inside the emulated amiga.
For 3.1 I could maybe adapt WHDLoad to use the physical installed one. But for 1.3 this cannot work as WHDLoad itself requires at least 2.0. Or could it be memory mapped somehow (other than the boot rom)?

We could insert it on a AmigaDOS disk drive e.g. df3 with only one binary file on root

**Reply by wepl** [reply]
this sounds fine for me
I could also handle if the kickstart is directly inserted into any DFx: (without any filesystem) by grabbing it there during the boot process.

**Comment by mithrendal** [comment]
I bumped into this kickstart list 

https://www.amigawiki.org/doku.php?id=de:system:overview_kickstart_versions

according to them there are some cbm kickstarts which have the same number, but are targeted for different CPUs 
 
e.g. 

```
Kickstart 2.04	37.175	390629-03	512kb	68020	1991	A3000, A3000T (Low)
Kickstart 2.04	37.175	390979-01	512kb	68000	1991	A500+, A2000, CDTV

Kickstart 3.1	40.63		512kb	68000	1994	A500, A600, A2000, CDTV
Kickstart 3.1	40.68		512kb	68020	1994	A3000, A3000T (Low)
```

is this relevant for our purposes ?

then prompting the user for just a kickstart version number is maybe not sufficient ?

e.g.
kickstart_rom_required:  [2.04, 3.1]

do we really need to specify more precisely?

kickstart_rom_required:  [ [2.04, 68020], [3.1, 68020]]

@wepl if you start with a 68000 CPU you can't do that with a 2.04 68020 anyways. 

maybe we ask always for the 68000 variants? On the other hand if user only has a 68020 rom ?

hm 🤔Any ideas?

**Comment by dirkwhoffmann** [comment]
My advice is not to overcomplicate things. Simply present a textual description to the user that this demo needs Kick 2.0 or whatever. When the user drags in a Rom file, merely start the emulator with that Rom. If the user has provided an invalid Rom, the demo won't work, like on a real Amiga with an incompatible Rom.

A more sophisticated approach would be to specify precisely which Roms are compatible. If you want to go down this road, my advice is not to distinguish Roms by name (you'll run into the problem you've already discovered yourself) but by their CRC32 fingerprints. Take the CRC checksum as the internal name of a Rom. vAmiga has a lot of functions for mapping CRC checksums back to user-friendly descriptions. For example, if you have a CRC number, you can get a string containing the Rom name, version number, etc.

**Reply by mithrendal** [reply]
>My advice is not to overcomplicate things.

yeah true... maybe everything higher than 2.04 is just okay for whdload and should work. So then I will continue to make this simple selection / save to local db of any cbm kickstart in order  to pass that into vAmigaWeb  ...

**Comment by wepl** [comment]
For WHDLoad installs there are only two versions relevant: 1.3 and 3.1.
For 1.3 there is only one version 34005.a500.
For 3.1 there exist 3 variants: 40063.a600 40068.a1200 40068.a4000. The first one works on 68000. The latter two only on 68020+.
Maybe it also makes sense that I explain the basic operation of WHDLoad here as I'm not sure you know already. WHDLoad completely takes over the system and runs the installed games in a degraded environment without any access to the Amiga host OS. Many games/demos do not use the OS, so when running under WHDLoad there is also no need for a kickstart image. But some games/demos use a lot of OS accesses. In this case WHDLoad uses a original kickstart image (kickemu). The image is loaded into the installed programs memory, is relocated to the specific memory address there and patched to use the WHDLoad API where applicable.
So to run WHDLoad I need any Kick 2.0+ and also AROS is ok.
For a subset of installs either a Kick 1.3 or Kick 3.1 is needed as file available at runtime.
Some very old install may need a specific 3.1 rom. All actual installs which require 3.1 are working with any 3.1. I would not make a difference here for now.
So there would be 3 kind of installs:
* no kickstart required
* kick 1.3 required
* kick 3.1 required

**Comment by mithrendal** [comment]
ok I have added now a new method to the player called `setup` ...

it is intended to be called right after when the complete html page is loaded into browser ... so that it can detect the kickstart requirements and display a lock / or the play button accordingly.

```javascript
<script>
        document.addEventListener('DOMContentLoaded', function() {
            vAmigaWeb_player.setup('teaser_pic', 
            {
                kickstart_rom_required:  [3.1],
                samesite_file: {
                    url:'./doc/media/tbl_eon_party.zip',name:'tbl.zip'
                },
                config:{
                    AROS:false,
                    wait_for_kickstart_injection:true,
                    navbar:false, 
                    wide:true,
                    startup_script: `df_mount_list=['tbl_eon_party_ver_disk1.adf']`
                }
            });
        });
</script>
```

The example above expects an element with id=`teaser_pic` in the html page and also another one that is an id like that plus appended with the suffix _overlay i.e. `teaser_pic_overlay`

```html
    <div id="container" style="position: relative;">
        <img id="teaser_pic" style="width:100%" src="https://vamiganet.github.io/footage/eon-small.jpg" />
        <div id="teaser_pic_overlay" style="position: absolute;top:0;left:0;width:100%;height:100%"></div>
    </div>
```


you can try out here 

https://vamigaweb.github.io/runcbm.html


ah yes ... dragging a rom file into it is not implemented yet ... currently only mouse click / touch to select rom file

when a cbm kickstart is loaded that way... it stays inside the local browsers database
<img width="1209" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/de3239a8-a686-40a8-b431-700c0a5fe694">

that way loading is only done once and the next time the player reads kickstart from this local database.

**Comment by mithrendal** [comment]
dragging a rom file into the overlay area is now also implemented 😎

**Comment by wepl** [comment]
thx, how can I access the kickstart image inside the emulation?

**Comment by mithrendal** [comment]
You mean the same kickstart with which the Amiga was started ? Can you not grab it from amiga memory ? I could write it otherwise to a floppy disk and then we could insert it in df0

**Reply by mithrendal** [reply]
@wepl not sure if this is what you need, but there is now the `mount_kickstart_in_dfn` parameter. The kickstart which the user has selected from local device (and stored in the local browser database) is not only injected into vAmigaWeb but also additionally imported as disk and mounted to DFn 

mount_kickstart_in_dfn: 0 

0 mounts it into df0, 1 into df1 and so on

```javascript
        document.addEventListener('DOMContentLoaded', function() {
            vAmigaWeb_player.setup('teaser_pic', 
            {
                kickstart_rom_required:  [1.3],
                samesite_file: {
                    url:'./doc/media/tbl_eon_party.zip',name:'tbl.zip',
                    mount_kickstart_in_dfn: 1
                },
                config:{
                    AROS:false,
                    wait_for_kickstart_injection:true,
                    navbar:false,
                    wide:true,
                    startup_script: `df_mount_list=['tbl_eon_party_ver_disk1.adf']`
                }
            });
        });
```

**Reply by mithrendal** [reply]
>So to run WHDLoad I need any Kick 2.0+ and also AROS is ok.
For a subset of installs either a Kick 1.3 or Kick 3.1 is needed as file available at runtime.

ok just I revisited what you explained above and I understand it in my own words as follows

1. to boot vAmiga with whdload AROS kickstart is ok
-> so we can go with setting `AROS:true`  and `wait_for_kickstart_injection:false` which boots vAmiga with the built in AROS kickstart

2. some installs need either 1.3 or 3.1 as a file (in whdload called the kickemu)
-> in this case we  go with
 `kickstart_rom_required: [1.3]` or `kickstart_rom_required: [3.1]` 
and `mount_kickstart_in_dfn: 1`





```javascript
        document.addEventListener('DOMContentLoaded', function() {
            vAmigaWeb_player.setup('teaser_pic', 
            {
                kickstart_rom_required:  [1.3],
                samesite_file: {
                    url:'./tbl_eon_party.zip',name:'tbl.zip',
                    mount_kickstart_in_dfn: 1
                },
                config:{
                    AROS:true,
                    wait_for_kickstart_injection:false,
                    navbar:false, 
                    wide:true,
                    startup_script: `df_mount_list=['tbl_eon_party_ver_disk1.adf']`
                }
            });
        });
```

this will boot vAmigaWeb with Aros kickstart and user selected kickstart as a file on a disk in df1:

**Comment by wepl** [comment]
thx a lot I will try to test this in the next days :-)

**Comment by wepl** [comment]
Did you maybe break something else?
The old setup is not working anymore it seems, e.g. https://whdload.de/ctros/BamigaSectorOne_GrandPrixC.html

**Reply by mithrendal** [reply]
hm that one brings this ...
<img width="695" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/d5a29a0d-5f16-432e-b5bd-17924b2a9a95">



but these others are still working fine

https://whdload.de/ctros/CopyrightDestroyersInc_3DT.html
https://whdload.de/ctros/Classic_ManchesterUtdEur.html
https://whdload.de/ctros/VisionFactory_5thGear.html
https://whdload.de/ctros/Trilogy_AmyPutt.html
https://whdload.de/ctros/CoolX_Pole500.html
https://whdload.de/ctros/TheCompany_AnotherWorld.html
...


seems they all work except the GrandPrixC, are you 100% sure the GrandPrixC was working before ?

**Reply by wepl** [reply]
Right, this one probably never worked.
I think it was a bad config on my other computer. It works, please forget my comment.

**Reply by mithrendal** [reply]
I have just made a small change...

the following three setup use cases should be working now

**1. specific kickstart for booting, kickemu on floppy drive**
- set a specific kickstart for booting with `kickstart_rom_url` `kickstart_ext_url` `AROS: false` `wait_for_kickstart_injection:true`
- set the user provided so called kickemu with  `kickstart_rom_required` and `mount_kickstart_in_dfn`
 
```javascript
   document.addEventListener('DOMContentLoaded', function() {
            vAmigaWeb_player.setup('teaser_pic', 
            {
                kickstart_rom_required:  [1.3],
                samesite_file: {
                    kickstart_rom_url:'./roms/aros-rom-20240313.bin',
	            kickstart_ext_url:'./roms/aros-ext-20240313.bin',
                    url:'./tbl_eon_party.zip',name:'tbl.zip',
                    mount_kickstart_in_dfn: 1
                },
                config:{
                    AROS:false,
                    wait_for_kickstart_injection:true,
                    // ...
                    startup_script: `df_mount_list=['tbl_eon_party_ver_disk1.adf']`
                }
            });
        });
```




**2. use built in AROS (if vAmigaWeb was not already preloaded with another kickstart) for booting, kickemu on floppy drive**

- set the fallback kickstart `AROS: true` `wait_for_kickstart_injection:false`
- set the so called kickemu with  `kickstart_rom_required` and `mount_kickstart_in_dfn`
 
this may be a bit unsafe because it might be that the user has preloaded vAmigaWeb with kickstart1.1 or emuTOS or something else completely weird. Safer is the first case i.e. use specific boot kickstart roms

```javascript
        document.addEventListener('DOMContentLoaded', function() {
            vAmigaWeb_player.setup('teaser_pic', 
            {
                kickstart_rom_required:  [1.3],
                samesite_file: {
                    url:'./tbl_eon_party.zip',name:'tbl.zip',
                    mount_kickstart_in_dfn: 1
                },
                config:{
                    AROS:true,
                    wait_for_kickstart_injection:false,
                    //..
                    startup_script: `df_mount_list=['tbl_eon_party_ver_disk1.adf']`
                }
            });
        });
```

**3. just prompt the user to select a specifc kickstart for booting e.g. kickstart 3.1, no kickemu required**

- set `AROS: false` and `wait_for_kickstart_injection:true`
- set the required commodore kickstart with  `kickstart_rom_required` 

```javascript
        document.addEventListener('DOMContentLoaded', function() {
            vAmigaWeb_player.setup('teaser_pic', 
            {
                kickstart_rom_required:  [3.1],
                samesite_file: {
                    url:'./tbl_eon_party.zip',name:'tbl.zip',
                },
                config:{
                    AROS:false,
                    wait_for_kickstart_injection:true,
                    //..
                    startup_script: `df_mount_list=['tbl_eon_party_ver_disk1.adf']`
                }
            });
        });

```

**Reply by wepl** [reply]
Sorry, no time yet to check this out. Will do if I found some time :)

**Reply by mithrendal** [reply]
Hi, no need to sorry 😀 ...

I fixed in the meantime a nasty thing which slowed down the loading when it  starts up. 😎

there was a sort of a race condition ...

on startup it run the Amiga async to the media hdf.zip loading...

which slowed down the whole loading, dezipping of the hdf.zip media. Escpecially when warp had been turned on then it had so little time left, that the preparation of the hdf volumne could take 3-5  seconds... you already saw the cat eyes of aros but the hdf was not yet inserted ...

today I synced the startup tasks ... now on startup, first thing load and decrunch the zip file and hdf volumne then after this start the Amiga Core... 

the result: it goes straight into the whdload screen ... no aros cat eyes anymore 😎

**Reply by wepl** [reply]
this is a great improvement!
I like to continue on this topic now and add to the kick13 support.
Sadly I'm stuck here at https://whdload.de/ctros/HighQualityCrackings_TD.html
It brings an 'Error: undefined'. I don't see what the problem is. Can you help?

**Reply by mithrendal** [reply]
> this is a great improvement! I like to continue on this topic now and add to the kick13 support. Sadly I'm stuck here at https://whdload.de/ctros/HighQualityCrackings_TD.html It brings an 'Error: undefined'. I don't see what the problem is. Can you help?

its the hard disk file ... I have put it into "vAmiga for Mac" version and here is what it says... 

<img width="260" alt="image" src="https://github.com/user-attachments/assets/351465a1-2467-4a2a-8c4d-4a1b871ecba4">

**Comment by mithrendal** [comment]
the eon demo which needs cbm kickstarts is now embedded in the show case room.

https://vamigaweb.github.io/doc/about.html

it is the last entry. 

... using the same  (**B**ring **Y**our **O**wn **K**ickstart) `BYOK` method ...

**Reply by wepl** [reply]
nice, but this demo install does not require a kickstart :)

**Reply by mithrendal** [reply]
> nice, but this demo install does not require a kickstart :)

it does 😎 the demo used for the BYOK (Bring Your Own Kickstart) use case (very last demo on about page) is the floppy disk file version of EON which needs original kickstarts. the whdload version although overall works fine with AROS, has some subtle timing  and skipped frames issues on two positions of the whole demo. already reported that to paraj / @mras0 . Did you know that paraj is also a contributor to vAmigaWeb where he implemented the hard disk driver among other things. Maybe he still remembers why no more than 16384 cyclinders for harddisk files are supported.

**Reply by wepl** [reply]
Ahh, I've overseen it. Thanks will check this as good example.

**Reply by mithrendal** [reply]
> Ahh, I've overseen it. Thanks will check this as good example.

I just pushed out version 2024_08_01 where I removed all limits for hdf geometry.

**Reply by wepl** [reply]
no need :)
this size was a bug, the hdf was much too large due a miscalc, was 250M

**Reply by mras0** [reply]
FWIW the HD code is only really limited to 4G per partition (though not really tested >= 2G), but vAmiga imposes additional constraints, in particular an absolute 504MB limit and less for "non-standard" sizes, i.e. ones that could not come from real drive geometry, since it tries to figure out a "true" C/H/S configuration.

The TBL-EON whdload version could probably be made to work better (though not 100% in sync) with some tweaking of settings. Like I wrote on [EAB](https://eab.abime.net/showpost.php?p=1680650&postcount=19) I think there is at least one part where their "up/down" loader only works because the floppy loader / decruncher routines are sufficiently slow on A500. There's also a similar issue with audio in a later part, so while it could probably be handled better, I don't really want to touch it :)

**Reply by wepl** [reply]
I don't understand how your html works. This too cryptic for me...

**Reply by wepl** [reply]
I have added 
`vAmigaWeb_player.kickstart_rom_required=[1.3];`
and
`mount_kickstart_in_dfn: 0`
here: https://whdload.de/ctros/HighQualityCrackings_TD.html
but there is no disk in df0:
you can call
`dic name=x`
to check

**Reply by mithrendal** [reply]
>I don't understand how your html works. This too cryptic for me...

because we can not do it like before ... 

the bring your own kickstart concept (BYOK) requires the html page to detect whether the user had already downloaded the kickstart possibly some days before.  If yes then the page shows a nice play icon and the user just can click on it. If no then the html page asks for the kickstart disk.

This detection on load of the html page makes a change necessary. We can not like we did before just execute some load script on the click event because we want the html UI to detect and show to the user on load whether the kickstart has already been stored before he clicks. For this reason we have to use the players new `setup` method as described above which does all that detection work for us when the html DOMloaded event comes in.

this evening after "real work" ;-) I will paste here the snippet which makes that together with

https://whdload.de/ctros/HighQualityCrackings_TD.html

.. stay tuned.

**Reply by mithrendal** [reply]
I have just copied the page to my domain and renamed it to

https://vamigaweb.github.io/doc/demo/whdload_byok.html

I had to change some of the URLs to make it work, (CORS browser policy would not allow my domain to cross resource sharing with the whdload.de site)


basically I had to insert at the beginning of the html file this piece of code, which becomes active when the html is fully loaded. And then displays to the user whether it already found a kickstart.

```javascript
<script>
	document.addEventListener('DOMContentLoaded', function() {
		vAmigaWeb_player.setup('HighQualityCrackings_TD', 
		{
			kickstart_rom_required:  [1.3],
			samesite_file: {
				kickstart_rom_url:'/roms/aros.bin',
				kickstart_ext_url:'/roms/aros_ext.bin',
				url:'./HighQualityCrackings_TD.zip',
				name:'HighQualityCrackings_TD.zip',
				mount_kickstart_in_dfn: 1
			},
			config:{
				AROS:false,
				wait_for_kickstart_injection:true,
				navbar:false,
				wide:true,
				startup_script: `
				hd_mount_list=['HighQualityCrackings_TD.hdf'];
				on_ready_to_run=()=>{
					wasm_configure('CPU_OVERCLOCKING','2');
					wasm_configure('CPU_REVISION','0');
					wasm_configure('AGNUS_REVISION','OCS');
					wasm_configure('DENISE_REVISION','OCS');
					wasm_configure('CHIP_RAM','512');
					wasm_configure('SLOW_RAM','0');
					wasm_configure('FAST_RAM','8192');
					wasm_configure('floppy_drive_count','1');
					wasm_configure('WARP_MODE','WARP_ALWAYS');
				}
				on_hdr_step=(drive_number,cylinder)=>{
					on_power_led_dim=()=>{
						wasm_configure('WARP_MODE','WARP_NEVER'); 
						on_power_led_dim=()=>{}; 
					}
					on_hdr_step=()=>{};
				}
				`
			}
		});
	});
</script>
```

the location where we had previously the loading code has also changed... since the code has been moved to the top there is no code anymore. Two additional elements are new now. 

```html
<b>Run emulated</b><br>
<div id="container" class="emulator_view" style="position: relative;width:50vw;aspect-ratio: 4 / 3;">
	<div id="HighQualityCrackings_TD" style="width:100%;"></div>
	<div id="HighQualityCrackings_TD_overlay" style="position: absolute;top:0;left:0;width:100%;height:100%;color:#eeeeee"></div>	
</div>	
```

maybe it would be wise to rename the ids "HighQualityCrackings_TD", "HighQualityCrackings_TD_overlay" to "va" and "va_overlay" to make it more generic


Also there was a bug in vAmigaWeb which made the option to mount the kickemu in df0 to fail. I fixed that bug. Then I started to realize that df0: is not a good option for the kickemu, because it is then always booting into df0 (with the kickemu inside) instead of dh0 😬.

so in the showcase above I choosed df1 for the kickemu floppy instead.



<img width="717" alt="image" src="https://github.com/user-attachments/assets/7c8e4996-38dc-4ee3-8489-1b56e7de7369">

going into the settings you can immediately check what drives have inserted disks ... in this case `dh0` and `df1`

**Reply by mithrendal** [reply]
🙈 floppy count was set to 1, therefore df1: was not visible to AmigaOS

```javascript
wasm_configure('floppy_drive_count','1');
```
<img width="734" alt="image" src="https://github.com/user-attachments/assets/33439de1-26fb-479a-943d-0848f377ebc5">

corrected that and here we see that the system sees the kickemu in df1:

**Reply by wepl** [reply]
Thanks a lot. It is working now.
https://whdload.de/ctros/HighQualityCrackings_TD.html

Could this approach to start va also used for stuff not require the kickstart image by omitting "kickstart_rom_required: [1.3]"?
This would simplify the html generation a bit and would look more similar.

Loading the kickstart from the floppy takes some time.
Is there any config option regarding floppy speed?

You provide the kickstart on a filesystem formatted disk. Initially I assumed that you simply put the kickstart image as raw adf into the drive. I would prefer this because it would speedup the copy process of the kickstart image. Reading without filesystem would be much faster. Maybe you can change this?

Using DF0: instead DF1: and having only one drive would save memory and speedup booting a bit. If there is no bootblock on the disk inserted it should still boot from the hdf. I don't see why we need 2 drives.

**Reply by mithrendal** [reply]
>Could this approach to start va also used for stuff not require the kickstart image by omitting "kickstart_rom_required: [1.3]"?

yes should work... I will test that

>Is there any config option regarding floppy speed?

```javascript
wasm_configure('DRIVE_SPEED', '8'); //1,2,4,8,-1
```

>Reading without filesystem would be much faster. Maybe you can change this?

this is what I currently do

```cpp
    // Create a new file system
    MutableFileSystem volume(INCH_35, DENSITY_DD, FS_OFS);
    volume.setName(FSName(filename));

    // Make the volume bootable
    volume.makeBootable(BB_AMIGADOS_13);
    
    // Add the file
    FSBlock *file = volume.createFile(filename, data.ptr, data.size);
    if (!file) throw VAError(ERROR_FS_OUT_OF_SPACE);
```

I will look into there what we have for options....

>If there is no bootblock on the disk inserted it should still boot from the hdf.

```cpp
    // Make the volume bootable
    volume.makeBootable(BB_AMIGADOS_13);
```

I should just remove these lines from current code and I guess we will be fine with df0: 😎

will do in the next days... stay tuned

**Reply by wepl** [reply]
> > Is there any config option regarding floppy speed?
> 
> ```js
> wasm_configure('DRIVE_SPEED', '8'); //1,2,4,8,-1
> ```

great 😄 

> > Reading without filesystem would be much faster. Maybe you can change this?
> 
> this is what I currently do
> 
> ```c++
>     // Create a new file system
>     MutableFileSystem volume(INCH_35, DENSITY_DD, FS_OFS);

FFS would also be a bit faster, for WHDLoad I need Kick 2+ anyway

> I should just remove these lines from current code and I guess we will be fine with df0: 😎
> 
> will do in the next days... stay tuned

👍

**Reply by mithrendal** [reply]
today I discovered the following API inside the memory component of vAmiga

```cpp
// Modifies Ram or Rom without causing side effects
    void patch(u32 addr, u8 *buf, isize len);
```

@dirkwhoffmann  just confirmed that it is for writing into amigas memory

@wepl what if we `patch` the kick-emu.rom into Amigas FAST-RAM instead of write it to a floppy disk ? Can you use it from there too ? That would be the fastest solution. 

Instead of 

```javascript
mount_kickstart_in_dfn: 0
```

we could set the following parameter which would write the kick.rom at the end of an 8MB fastram expansion

```javascript
patch_kickstart_into_address: 0x00980000
```

**Reply by wepl** [reply]
Yes, I can handle this. I can write a small tool which grabs the kick from there.
Would be the fastest solution 😃

**Reply by mithrendal** [reply]
very cool !

published version 2024_08_08 which ist now able to patch amiga memory from javascript 😎

I tested it on my testpage  https://vamigaweb.github.io/doc/demo/whdload_byok.html with parameter

```javascript
patch_kickstart_into_address: 0x00980000
```

and that logs into the javascript console the following output 

```
wasm_mem_patch addr=0x980000, len=524288, header bytes= 11, 11, 4e
```

I also added a black color and border to the container element because I thought that this looks nicer
```css
background-color: black;border-radius: 15px;padding: 15px;
```

**Reply by wepl** [reply]
Thanks!
Testing now.
It seems writing to the hdf provided seems to be broken somehow. Is this known or not supported?
There is no error, but the content of the written file is all zeros.

**Reply by wepl** [reply]
https://whdload.de/ctros/HighQualityCrackings_TD.html
<img width="861" alt="Bildschirmfoto 2024-08-09 um 00 21 28" src="https://github.com/user-attachments/assets/5871b2b9-2feb-451a-8ee4-58a2c98b0b62">

**Reply by mithrendal** [reply]
>It seems writing to the hdf provided seems to be broken somehow. Is this known or not supported?
There is no error, but the content of the written file is all zeros.

I think hd writing is fine. The problem is that after startup and kickemu mem patch we make several changes to configuration of memory, connected hard drive and so on. Some of these subsequent configurations will power down and power on the amiga machine. So that the initial mem patch which was in transient fast memory only is lost due to the power down of the memory chips. Hence kickemu patch was reverted and your savemem saved only zeros.

just did a quick temporary hack... that does the kickemu patch after the step motor of dh0: moves.  Which shows the expected result ...

<img width="628" alt="image" src="https://github.com/user-attachments/assets/e7fcc652-75e4-4f9b-8432-7a11bf76c719">



I published that quick hack but did not increment version number yet. So to try out you have to delete current version 2024_08_08 via settings and reinstall.

**Reply by wepl** [reply]
works fine now 😄

**Reply by wepl** [reply]
> I also added a black color and border to the container element because I thought that this looks nicer
> 
> ```css
> background-color: black;border-radius: 15px;padding: 15px;
> ```

Yes, looks better. Have taken this.

**Reply by wepl** [reply]
> > Could this approach to start va also used for stuff not require the kickstart image by omitting "kickstart_rom_required: [1.3]"?
> 
> yes should work... I will test that

Doesn't seem to work: https://whdload.de/ctros/CopyrightDestroyersInc_3DT.html

**Reply by mithrendal** [reply]
>doesn't seem to work

🙈 true.  when I did this I had only this user provided kickemu feature in mind. Did not test the normal use case where we don't need an extra cbm kickstart. I will do this the following days, so that we also can use the players setup method for this. Expect it somewhen in the middle of the coming week.


>Yes, looks better. Have taken this.

I think aspect ratio 3 / 2 looks even better on the container. Also because the bottom border is not bigger than the other edges when player is spawned. What do you think?

```html
<div id="container" class="emulator_view" 
style="position:relative;width:50vw;aspect-ratio:3/2;background-color:black;
border-radius:15px;padding:15px">
```

the https://whdload.de/ctros/HighQualityCrackings_TD.html is already working with the user supplied kickemu right ? when I click on it the demo runs well. Very cool !!

**Reply by mithrendal** [reply]
>Doesn't seem to work: https://whdload.de/ctros/CopyrightDestroyersInc_3DT.html

fixed it. both demos work now. the one which requires the kickemu and the one above which doesn't. 😎

@wepl  I did not yet increment the version. if you think its good I will increment it so that an update notification comes up. therefore to get the new version you have to reinstall vAmigaWeb via settings (or use incognito / private browsing)

**Reply by wepl** [reply]
Works great now 👍 Thanks a lot!
I changed the aspect to 3/2, looks better.
No need to increment the version from my side.
I will now fix the ctros not working ATM and then start to provide images for all the mags...

**Reply by mithrendal** [reply]
>I will now fix the ctros not working ATM and then start to provide images for all the mags...

cool. I don't know which images you got but when you have pictures of these ctros then you could also replace the `id="va"` element which displays the black box

```html
<div id="va" style="width:100%"></div>
```

with an teaser image of the crto like this

```html
<img id="va" style="width:100%" src="https://vAmigaWeb.github.io/doc/images/batman_rises.jpg">
```

which would makes it look like this

<img width="741" alt="image" src="https://github.com/user-attachments/assets/62426967-512f-4d5f-b37f-370f006f37d0">

**Reply by wepl** [reply]
thanks for the tip, but for now I have no time to manage such pictures

**Comment by mithrendal** [comment]
I just discovered that patching the kickemu into fastram  `patch_kickstart_into_address:0x00980000` currently seems to fail with vAmigaWeb v3... it was working fine with v2.6

starting the following cracktro where this technique is in use
https://www.whdload.de/ctros/UnitA_FA18Interceptor.html


with vAmigaWeb v3 leads to
![image](https://github.com/user-attachments/assets/e35ea6cb-b02c-474b-94d0-67adb7097d9b)

with vAmigaWeb v2.6 it works still fine. (tested by downgrading via settings)

@wepl I assume vAmigaWeb isn't applying the emukick patch to fastram correctly... Could that be the problem? Are you able to interpret the error message?

**Reply by wepl** [reply]
works here, do you use a different kickstart?

**Reply by wepl** [reply]
the error message is from the aros rom, I assume before whdload is started

**Reply by mithrendal** [reply]
I plugged in a kick1.3 ... which works when I downgrade to vAmigaWeb v2.6

>works here ...

maybe you are still on v2.6 (you can check in settings)

>I assume before whdload is started

the error comes when hard drive begins to step its motor... which is the time when we also start to patch the emukick into FASTRAM  0x00980000

**Reply by wepl** [reply]
I have running the latest version, core 3.0.1, ui 2025_01_08.

**Reply by wepl** [reply]
I have installed a debug hdf.
Can you retry?

**Reply by mithrendal** [reply]
when I use my kick 1.3 rom then it still does not work.

when I plug a kick1.2 into it then this

<img width="610" alt="image" src="https://github.com/user-attachments/assets/6017448e-472c-4c1c-afc7-2f7d40e75376" />

I don't get it why it works with the same kick1.3 in vAmigaWeb v2.6 and not in vAmigaWeb 3.0.1 ... even more strange that it works on your site

**Reply by wepl** [reply]
Has the kickstart 1.3 the correct size (256K)?
With the 1.2 rom, whdload should complain when you start it via 'whdload * data data'

**Reply by mithrendal** [reply]
>Has the kickstart 1.3 the correct size (256K)?

ah no it has 512k. Is that a problem ?

**Reply by wepl** [reply]
It should not be a problem. But it's still a good idea to fix it. Starting with 2.0 the ROM is 512K, prior they are 256K.
The crash seems somewhere in the AROS boot process. No idea.
Did you try a CBM ROM instead the AROS one?

**Reply by mithrendal** [reply]
problem has been solved in issue ticket

https://github.com/vAmigaWeb/vAmigaWeb/issues/276
